### PR TITLE
[16.0][MIG] product_variant_configurator: Migration to 16.0

### PR DESCRIPTION
**Pull request by SergioPSV**
Standard Migration

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Thanks for the contribution.

Please preserve commit history following technical method explained in https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0.

If the jump is between several versions, you have to modify the source branch in the main command to accommodate it to this circumstance.

**Comment by SergioPSV** [comment]
@OCA/core-maintainers

**Comment by SergioPSV** [comment]
@mariadforgeflow Can you review the code?

**Comment by pedrobaeza** [comment]
Please see my previous comment: https://github.com/OCA/product-variant/pull/286#issuecomment-1453996944

**Comment by SergioPSV** [comment]
@pedrobaeza Added the history of commits

File: product_variant_configurator/models/product_configurator.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        compute="_compute_price_extra",
+        digits="Product Price",
+        help="Price Extra: Extra price for the variant with the currently "
+        "selected attributes values on sale price. eg. 200 price extra, "
+        "1000 + 200 = 1200.",
+    )
+    product_id = fields.Many2one(
+        string="Product Variant", comodel_name="product.product"
+    )
+    can_create_product = fields.Boolean(compute="_compute_can_be_created", store=False)
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:37]: should remove store=False while it is default.

File: product_variant_configurator/models/product_configurator.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    create_product_variant = fields.Boolean(string="Create product now!")
+
+    @api.depends(
+        "product_attribute_ids", "product_attribute_ids.value_id", "product_id"
+    )
+    def _compute_can_be_created(self):
+        for rec in self:
+            if rec.product_id:
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:45]: we can do like `if rec.product_id or not rec.product_tmpl_id`:

File: product_variant_configurator/models/product_configurator.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                # no product nor template
+                rec.can_create_product = False
+                continue
+            rec.can_create_product = not bool(
+                len(rec.product_tmpl_id.attribute_line_ids.mapped("attribute_id"))
+                - len(list(filter(None, rec.product_attribute_ids.mapped("value_id"))))
+            )
+
+    @api.depends("product_attribute_ids", "product_attribute_ids.value_id")
+    def _compute_price_extra(self):
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:59]: i think we can re-use function `_compute_can_be_created`

File: product_variant_configurator/models/product_configurator.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        for record in self:
+            record.price_extra = sum(record.mapped("product_attribute_ids.price_extra"))
+
+    def _set_product_tmpl_attributes(self):
+        if self.product_tmpl_id:
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:64]: should add `self.ensure_one()`?

File: product_variant_configurator/models/product_configurator.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                        "attribute_id": attribute_line.attribute_id.ids[0],
+                        "product_tmpl_id": self.product_tmpl_id.ids[0],
+                        "owner_model": self._name,
+                        "owner_id": self.id,
+                    }
+                )
+            self.product_attribute_ids = attribute_lines
+
+    def _set_product_attributes(self):
+        if self.product_id:
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:78]: should use `self.ensure_one()`?

File: product_variant_configurator/models/product_configurator.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        self.product_attribute_ids = self.product_attribute_ids.browse([])
+
+    @api.onchange("product_tmpl_id")
+    def _onchange_product_tmpl_id_configurator(self):
+        self.ensure_one()
+        if not self.product_tmpl_id._origin:
+            self.product_id = False
+            self._empty_attributes()
+            # no product template: allow any product
+            return {"domain": {"product_id": []}}
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:97]: i think this doesn't work anymore from version 15. Use onchange to set the domain.

File: product_variant_configurator/models/product_configurator.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+        if not self.product_tmpl_id.attribute_line_ids:
+            # template without attribute, use the unique variant
+            self.product_id = self.product_tmpl_id.product_variant_ids[0].id
+        else:
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:102]: `elif self.product_id and self.product_id.product_tmpl_id != self.product_tmpl_id and not self.env.context.get("not_reset_product") =>                     self.product_id = False`

File: product_variant_configurator/models/product_configurator.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        if self.product_id:
+            self._set_product_attributes()
+        elif self.product_tmpl_id:
+            self._set_product_tmpl_attributes()
+        else:
+            self._empty_attributes()
+
+        # Restrict product possible values to current selection
+        domain = [("product_tmpl_id", "=", self.product_tmpl_id.ids[0])]
+        return {"domain": {"product_id": domain}}
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:122]: i think this doesn't work anymore.

File: product_variant_configurator/models/product_configurator.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+    @api.onchange("product_attribute_ids")
+    def _onchange_product_attribute_ids_configurator(self):
+        self.ensure_one()
+        if not self.product_tmpl_id:
+            return {"domain": {"product_id": []}}
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:128]: same above.

File: product_variant_configurator/models/product_configurator.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        if not self.product_attribute_ids:
+            domain = [("product_tmpl_id", "=", self.product_tmpl_id.id)]
+            return {"domain": {"product_id": domain}}
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:131]: same above.

File: product_variant_configurator/models/product_configurator.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                obj = self.env["product.attribute.value"].with_context(
+                    lang=self.partner_id.lang
+                )
+                values = obj.browse(self.product_attribute_ids.mapped("value_id").ids)
+                obj = self.env["product.template"].with_context(
+                    lang=self.partner_id.lang
+                )
+                product_tmpl = obj.browse(self.product_tmpl_id.id)
+            self.name = self._get_product_description(product_tmpl, False, values)
+        return {"domain": {"product_id": domain}}
```
Comment (hieulucky111) [inline product_variant_configurator/models/product_configurator.py:158]: same above.

**Comment by ivs-cetmix** [comment]
Hey @SergioPSV thank you for your contribution! 
Looks like runboat is down. Could you please push something to this PR to bring it up again?
We will do the functional testing

**Comment by SergioPSV** [comment]
@isserver1 runboat restart successfully

**Review by hieulucky111 — CHANGES_REQUESTED** [review]

**Comment by maxgrassi** [comment]
Hi, I tested  and it works great! I have a suggestion: Exclude/Filter variant for attribute combination. Because the standard behavior in the Sales app (sales order) or ecommerce app (product page) when configuring attributes and variants it always shows all product configurations as set on the master product (product.template) and does not filter products that actually exist at the product.product variant level.

When configuring product attributes and variants the system does not filter the available combinations but lists all combinations even those that do not exist. This is quite confusing  for example in complex configurable products.

**Comment by LoisRForgeFlow** [comment]
@SergioPSV Could you attend the last two reviews? Thanks!

**Comment by EvaSForgeFlow** [comment]
Continuation of this PR: https://github.com/OCA/product-variant/pull/303

**Comment by LoisRForgeFlow** [comment]
> Hi, I tested and it works great! I have a suggestion: Exclude/Filter variant for attribute combination. Because the standard behavior in the Sales app (sales order) or ecommerce app (product page) when configuring attributes and variants it always shows all product configurations as set on the master product (product.template) and does not filter products that actually exist at the product.product variant level.
> 
> When configuring product attributes and variants the system does not filter the available combinations but lists all combinations even those that do not exist. This is quite confusing for example in complex configurable products.

I think there could be performance issues for products with several attributes in which the combination of all them is really huge. So I see such a feature out of the scope of the migration, it could be added in a separate PR and discuss the possibilities there.

**Comment by hildickethan** [comment]
You can close this as https://github.com/OCA/product-variant/pull/303 is merged

**Pull request closed without merging**